### ISSUE TITLE
Include full sharpness levels from mhwg.org

### ISSUE THREAD

**Issue by stocky37** [LartTyler/MHWDB-API#51]
Original question: Do you have any plans to parse mhwg.org for handicraft sharpness levels? If not, could you? :)

**Comment by LartTyler** [comment]
To clarify, you're referring to the values in the second to last column on this page (as an example): http://mhwg.org/data/4000.html. Correct? If so, I'll get it added to the backlog.

**Comment by stocky37** [comment]
I believe those values just show the min and max sharpness values, but not each individual level. If you head down into the individual weapon pages (e.g. [fire-and-ice](http://mhwg.org/ida/220822.html)) you'll notice the third column in the main table contains a bar for each handicraft level.

**Comment by LartTyler** [comment]
Aha! I understand now. Lemme think on this one, since they use a different scale for their bars (MWHDB represents each sharpness as a percentage of a max 100, theirs seems to cap at 160), but I think it's definitely worth representing somehow.

**Comment by stocky37** [comment]
Awesome, cheers. No rush since I can work around it manually for the weapons I actually care about for now, but would definitely pretty sweet to have long term. For context I'm hoping to show the sharpness related to the handicraft skill level, so when making builds you can see how much sharpness you're actually getting.

**Comment by bolte-17** [comment]
I suspect mhwg is more accurate. The true scale is also shown a bit on mhwg- each rank of handicraft adds 10 units of sharpness in game terms (good for 10 hits normally). I believe each bar section is also a multiple of 10- 1px differences on kiranico's bars are just an artifact.

The mhwg bars are 160px wide instead of 100 and seem to be at a constant 4px : 10 sharpness ratio. So something like Anguish (Jho sword) at Handicraft 5 would have true sharpness values of:
```
{
    red: 60,
    orange: 120,
    yellow: 80,
    green: 60,
    blue: 30,
    white: 50
}
```

It looks like there's a limit of 400 total sharpness for any given weapon. 

If the API provides all the sharpness breakpoints, you'd only need one additional data point of "starting total sharpness" to fully model the sharpness/handicraft interaction.

**Comment by LartTyler** [comment]
@bolte-17 I see what you're saying, thanks for the info. However, there's one thing I'm not quite wrapping my head around, but maybe I'm just missing something.. How would adding just starting total sharpness provide all the information you'd need? Wouldn't I need to store sharpness at each handicraft level? As far as I can tell, there's no rhyme or reason to when handicraft stops increasing the weapon's default max sharpness color, and moves into the next tier (i.e. when [Chrome Razor 2](http://mhwg.org/ida/220822.html) stops increasing blue sharpness and adds white sharpness). Is there some formula to it, or does it vary from weapon to weapon?

**Comment by stocky37** [comment]
I think @bolte-17 is on to something. Assuming a weapon has 400 total sharpness (the number doesn’t really matter for the example), then it looks like each particular weapon has “breakpoints” if you will for when it reaches a particular level of sharpness, and those breakpoints are the same no matter what level of handicraft.

So say we have a weapon that has the following breakpoints:
‘’’
{
  “red”: 0,
  “orange”: 100,
  “yellow”: 150,
  “green”: 250,
  “blue”: 320,
  “white”: 370
‘’’

These numbers will always be the same for all levels of handicraft, but, assuming this weapon will reach max sharpness with 5 levels of handicraft, it would start with a sharpness value of 350, and therefore start off with approximately 30 blue sharpness. Add a level of handicraft and you will have 40 etc. Then, when you hit level 3 of handicraft, you will get 10 white sharpness.

**Comment by bolte-17** [comment]
It looks like most fully-upgraded weapons do fill the bar with handicraft 5, exceptions being weapons like Nerg and Odo's which naturally have full bar and do not benefit from handicraft.

I don't believe handicraft 5 will fill the bar fully on unupgraded weapons, though.

**Comment by bolte-17** [comment]
Ah, I think I understand the confusion now. You would need the starting total sharpness value in addition to either the sharpness breakpoints (@stocky37's example) or the sharpness-per-tier values (My Jho GS example)

**Comment by bolte-17** [comment]
I think I like the breakpoints format more. Easier to work with to answer "what tier am I in" and "how much handicraft to get the next tier"

**Comment by TanukiSharp** [comment]
Is there a way to help ? I'm not found of PHP but I could make a tool to massively extract sharpness data from mhwg.org and produce a JSON for you (@LartTyler) to easily integrate in the DB.

**Comment by stocky37** [comment]
> I don't believe handicraft 5 will fill the bar fully on unupgraded weapons, though.

I haven't seen one that does.

> I think I like the breakpoints format more. Easier to work with to answer "what tier am I in" and "how much handicraft to get the next tier"

That's pretty much exactly what I was going for. Both would work perfectly fine but breakpoints should mean less calculations required for most use cases.

**Comment by LartTyler** [comment]
I think the breakpoints idea is gonna be the way to go. @TanukiSharp, as far as helping out goes, a JSON dump of the breakpoints would be awesome. If you plan on creating a tool to do so, I'm sure there's some way we could incorporate it into the API's scraping process (i.e. a script that compiles the data into a JSON file, which can then be imported during scraping).

**Comment by TanukiSharp** [comment]
What is the "API scraping process" ?

I was thinking about creating a small C# tool (and put the code on GitHub) that queries the HTML pages, parses stuffs, get images, extracts values and outputs usable data. This process might be a bit long, requests intensive and efficiency will heavily rely on availability of the mhwg.org hosting service, so something that better be done offline.

Anyway, I will start soon and let you know my progress. I just want to make sure we are on the same page so I will take the bellow example, the `Chrome Razor 2`, value is out of 160:

Values set A
```
24  red
24  orange
 8  yellow
60  green
36  blue
 8  white
```

Values set B
```
 24  red
 48  orange
 56  yellow
116  green
152  blue
160  white
```

Which values set do you prefer ?

**Comment by LartTyler** [comment]
The "scraping process" that I'm referring to is a bunch of code that scrapes data from places like Kiranico and Fextralife. This is done periodically and at release time on a dedicated staging machine, at which point a new database snapshot is taken and stored in the `/snapshots` directory, which is then loaded into the production database.

As far as structure goes, option B is gonna be the best choice.

**Comment by TanukiSharp** [comment]
Got it. Last question, do you prefer me to put the tool in my repo and you can reference it, or should I clone this repo, put the tool there and make a PR ?

**Comment by LartTyler** [comment]
In it's own repository would be best. Thanks!

**Comment by TanukiSharp** [comment]
Quick report, I'm making good progress on the tool, I think I can get a prototype working by tomorrow at best, this week-end at worst.

Since mhwg.org has weapon names in Japanese only and I don't want to maintain a mapping between mhwg.org weapons names and mhw-db.com weapon ids, I parse as many info as I can from mhwg.org and fetch weapons from mhw-db.com at the same time, and then automatically match the weapons based on as many arguments as I can, so then I can update the sharpness of mhw-db.com weapons from the correct one of mhwg.org.

Technically, this tool could do more than merging sharpness info, but for the moment I will not go further than that.

I will keep you posted.

**Comment by stocky37** [comment]
If we're going with value set B (the breakpoints), you've got the values wrong. The numbers there are supposed to represent where the sharpness levels *start*. So the example out of 160 should be:
```
0  red
24  orange
48  yellow
56  green
116  blue
152  white
```
I think we should also be storing the sharpness values out of 400 so we can do direct calculations rather than out of 160. We can just multiply by 2.5, so given the above we'd end up with:
```
0  red
60  orange
120  yellow
140  green
290  blue
380  white
```

**Comment by TanukiSharp** [comment]
> The numbers are supposed to represent where the sharpness levels start.
That makes 2 questions to answer then:
1) Why storing the red, since it will always start to 0
2) How do you know where the white ends ?

**Comment by stocky37** [comment]
Good point, you wouldnt need to store red.
As for the second question, I’ve only been thinking about final weapons so it hadn’t occurred to me that weapons wouldnt fill out to max sharpness. Maybe we do need to store where each sharpness level finishes instead.

**Comment by TanukiSharp** [comment]
Yes, so I guess I don't need to change anything, just apply a multiplier, out of 400 is fine to me. I think I never saw two data sources giving the same standard max range, so whatever value will do.

**Comment by TanukiSharp** [comment]
The base tool to fetch mhwg.org is done. As I said, it also fetches weapons data from mhw-db.com in order to find matches. Here is a very quick report:

- mhwg.org contains 974 sharpness-aware weapons
- mhw-db.com contains 935 sharpness-aware weapons (-39)
- there are only 662 matching weapons between the two data sources (intersect)

I will try to give you such report per weapon type tonight or this week-end.

A question that you can answer before anyway, what kind of output format do you want ? Something like the output of the HTTP API ? Or do you want a specific format ?

Thanks.

**Comment by TanukiSharp** [comment]
Actually I had time to write a quick query, and here is the results split per weapon type:

| Weapon type | mhwg.org | mhw-db.com | Diff | Intersect |
| --- | --- | --- | --- | --- |
| GreatSword | 95 | 92 | -3 | 67 |
| LongSword | 87 | 83 | -4 | 62 |
| SwordAndShield | 97 | 94 | -3 | 70 |
| DualBlades | 88 | 85 | -3 | 64 |
| Hammer | 89 | 86 | -3 | 66 |
| HuntingHorn | 90 | 87 | -3 | 0 |
| Lance | 90 | 86 | -4 | 66 |
| Gunlance | 85 | 81 | -4 | 62 |
| SwitchAxe | 85 | 81 | -4 | 66 |
| ChargeBlade | 82 | 77 | -5 | 58 |
| InsectGlaive | 86 | 83 | -3 | 81 |

I found out that Hunting Horns do not have info about melodies / notes, but even without taking this in account for matching, I still get the same result, I have to figure out what is going on, probably a bug in my parsing of mhwg.org.

**Comment by TanukiSharp** [comment]
OK, I found some bugs in my code, fixed them and now reached 842 matching weapons between the two data sources.

Anyway, I also found a way to easily found mistakes in the data, so I will run this in parallel with the game to find out from which data source the mistakes come from, and will post again. (this week-end)

**Comment by LartTyler** [comment]
@TanukiSharp As far as output is concerned, a JSON array containing objects that include the MHWDB weapon IDs and the sharpness breakpoints would probably be the easiest to work with. But any output should be fine, so long as it's clear what all the values are.

**Comment by TanukiSharp** [comment]
OK for the output format.

Here is the list of all the differences I found. I checked all those difference one by one manually with the game in order to confirm they are real mistakes. In the process, I found 3 mistakes in mhwg.org which I will report too right away.

Do you mind fixing them all before I can proceed to extract the sharpness ?

Thanks a lot.

---

**GreatSword**

Buster Blade 3 (id: 6)
- slot should be '1--' but is '---'

Jagras Hacker 1 (id: 23)
- element 1 should be '(Water 210)' but is '(Ice 210)'

Jagras Hacker 2 (id: 24)
- element 1 should be '(Water 240)' but is '(Ice 240)'

Jagras Hacker 3 (id: 25)
- element 1 should be '(Water 300)' but is '(Ice 300)'

Gnashing Flammenzahn (id: 56)
- element 1 should be 'Fire 480' but is 'Fire 360'

Daora's Decimator (id: 85)
- affinity should be 10 but is 0

Wyvern Ignition "Impact" (id: 1181)
- element 1 should be '(Fire 510)' but is '(Fire 240)'

Missing weapon ガイラブレイズ・水
Missing weapon ガイラブレイズ・氷
Missing weapon ガイラブレイズ・睡眠

---

**LongSword**

Brazenridge 2 (id: 111)
- defense should be 30 but is 25

Missing weapon: Azure Star "Dragon Dance"
- rarity: 7
- attack: 495
- affinity: 30%
- defense: 0
- element: 'Water 180'
- slots: '3--'
- craft: ?
- cost: ?
- augmentation slots: 2

Missing weapon ガイラソード・水
Missing weapon ガイラソード・毒
Missing weapon ガイラソード・麻痺

---

**SwordAndShield**

Lumu Tabar 3 (id: 205)
- slot should be '31-' but is '21-'

Grand Barong 2 (id: 224)
- defense should be 10 but is 0

Carapace Edge 1 (id: 241)
- affinity should be -10 but is 0

Carapace Edge 2 (id: 242)
- affinity should be -10 but is 0

Carapace Edge 3 (id: 243)
- affinity should be -10 but is 0

Barroth Club 3 (id: 246)
- slot should be '21-' but is '11-'

Missing weapon ガイラスライサー・水
Missing weapon ガイラスライサー・龍
Missing weapon ガイラスライサー・爆破

---

**DualBlades**

Freeze Daggers 2 (id: 280)
- element 1 should be 'Ice 90' but is 'Ice 60'

Diablos Clubs 2 (id: 308)
- defense should be 15 but is 10

Dragonbone Twinblades 2 (id: 333)
- element 1 should be 'Dragon 150' but is 'Dragon 120'

Twin Nails (id: 342)
- affinity should be 10 but is 0

Fire and Ice (id: 343)
- affinity should be 10 but is 0

Xeno Raqs (id: 344)
- affinity should be 15 but is 0

Missing weapon ガイラダガー・水
Missing weapon ガイラダガー・氷
Missing weapon ガイラダガー・毒

---

**Hammer**

Lightning Bash 1 (id: 376)
- element 1 should be 'Thunder 240' but is 'Thunder 832'

Diablos Sledge 2 (id: 403)
- element 1 should be '(Ice 150)' but is '(Ice 120)'

Anja Striker 3 (id: 416)
- element 1 should be 'Fire 450' but is 'Fire 390'

Dragonbone Basher 3 (id: 419)
- slot should be '1--' but is '---'

Icesteel Hammer (id: 424)
- affinity should be 10 but is 0

Daora's Colossus (id: 425)
- affinity should be 10 but is 0

Xeno Maph'agarna (id: 426)
- affinity should be 15 but is 10

Devil's Due (id: 1158)
- slot should be '---' but is '3--'

Devil's Crush (id: 1159)
- slot should be '---' but is '3--'

Missing weapon ガイラハンマー・水
Missing weapon ガイラハンマー・雷
Missing weapon ガイラハンマー・氷

---

**HuntingHorn**

Gama Horn 1 (id: 472)
- slot should be '22-' but is '2--'

Gama Horn 2 (id: 473)
- slot should be '22-' but is '2--'

Lumu Barone 1 (id: 487)
- element 1 should be '(Water 420)' but is 'Water 420'

Lumu Barone 2 (id: 488)
- element 1 should be '(Water 480)' but is 'Water 480'

Lumu Barone 3 (id: 489)
- element 1 should be '(Water 570)' but is 'Water 570'

Missing weapon ガイラバイプ・水
Missing weapon ガイラバイプ・龍
Missing weapon ガイラパイプ・麻痺

---

**Lance**

Blue Tail (id: 550)
- element 1 should be 'Fire 270' but is 'Fire 210'

Garon Lance 2 (id: 574)
- affinity should be 15 but is 0

Dragonbone Lance 3 (id: 585)
- slot should be '11-' but is '1--'

Thunderpike (id: 588)
- affinity should be 10 but is 0

Missing weapon ガイラクレスト・水
Missing weapon ガイラクレスト・雷
Missing weapon ガイラクレスト・睡眠
Missing weapon ガイラクレスト・惨爪

---

**Gunlance**

Steel Gunlance 3 (id: 603)
- has no element but should be '(Blast 210)'

Jyura Buster 1 (id: 626)
- slot should be '1--' but is '---'

Jyura Buster 2 (id: 627)
- slot should be '1--' but is '---'

Jyura Buster 3 (id: 628)
- slot should be '1--' but is '---'

Missing weapon ガイラバスター・水
Missing weapon ガイラバスター・毒
Missing weapon ガイラバスター・睡眠
Missing weapon ガイラバスター・王

---

**SwitchAxe**

Power Smasher 1 (id: 716)
- elderseal should be Average but is None

Power Smasher 2 (id: 717)
- slot should be '3--' but is '---'
- elderseal should be Average but is None

Dragonbone Slicer 1 (id: 745)
- elderseal should be Low but is Average

Dragonbone Slicer 2 (id: 746)
- elderseal should be Low but is Average

Dragonbone Slicer 3 (id: 747)
- elderseal should be Low but is Average

Xeno Martshu (id: 755)
- affinity should be 15 but is 0

Missing weapon ガイラアックス・雷
Missing weapon ガイラアックス・睡眠
Missing weapon ガイラアックス・麻痺
Missing weapon ガイラアックス・角

---

**ChargeBlade**

Hard Bone Strongarm 3 (id: 789)
- slot should be '1--' but is '---'

Missing weapon: Dante's Devil Sword
- rarity: 8
- attack: 720
- affinity: 0%
- defense: 0
- element: 'Thunder 150'
- slots: '11-'
- phial type: Impact
- craft: Red Orb x2, Anjanath Gem x1, Odogaron Gem x1, Teostra Mane x4
- cost: 24000z
- augmentation slots: 1

Missing weapon ガイラアームズ・雷
Missing weapon ガイラアームズ・氷
Missing weapon ガイラアームズ・睡眠
Missing weapon ガイラアームズ・角

---

**InsectGlaive**

Nergal Reaper (id: 839)
- elderseal should be High but is None

Catastrophe's Light (id: 840)
- elderseal should be High but is None

Datura Blade 1 (id: 863)
- slot should be '11-' but is '---'

Hazak Entoma 1 (id: 876)
- elderseal should be Average but is None

Hazak Entoma 2 (id: 877)
- elderseal should be Average but is None

Garon Rod 2 (id: 898)
- slot should be '1--' but is '---'

Dragonbone Glaive 1 (id: 901)
- elderseal should be Low but is None

Dragonbone Glaive 2 (id: 902)
- elderseal should be Low but is None

Dragonbone Glaive 3 (id: 903)
- elderseal should be Low but is None

Blacksteel Glaive 1 (id: 907)
- elderseal should be Low but is None

Blacksteel Glaive 2 (id: 908)
- elderseal should be Low but is None

Xeno Shmaena (id: 911)
- elderseal should be Low but is None

Bad Wing (id: 1170)
- elderseal should be High but is None

Grunge Storm (id: 1171)
- elderseal should be High but is None

Missing weapon ガイラブレード・水
Missing weapon ガイラブレード・氷
Missing weapon ガイラブレード・麻痺

**Comment by TanukiSharp** [comment]
Here is a first prototype that "works on my machine": https://github.com/TanukiSharp/MHWSharpnessExtractor
Let me know if you need help to make it work, or if you prefer me to generate the sharpness JSON file for you.

**Comment by LartTyler** [comment]
@TanukiSharp It'll take me a while to get through all of that incorrect data, but it definitely needs to be fixed. Thank you for taking the time to find all of that. I'm going to open a separate issue dedicated to resolving those errors, as it's a bit of a bigger issue, as most of that information is being pulled from Kiranico, and needs to come from a different source if it's that wrong.

**Comment by LartTyler** [comment]
Full sharpness values are now available in the API in [v1.13.0](https://github.com/LartTyler/MHWDB-API/releases/tag/1.13.0).
